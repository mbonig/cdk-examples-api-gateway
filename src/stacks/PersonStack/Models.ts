import { IdentifiableModel } from '../common/handlers/IdentifiableModel';

export interface Person extends IdentifiableModel {
  firstName: string;
  lastName: string;
}

