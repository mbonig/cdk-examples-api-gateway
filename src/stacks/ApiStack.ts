import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Deployment, RestApi, Stage } from 'aws-cdk-lib/aws-apigateway';
import { Certificate, CertificateValidation } from 'aws-cdk-lib/aws-certificatemanager';
import { AttributeType, TableV2 } from 'aws-cdk-lib/aws-dynamodb';
import { ARecord, HostedZone } from 'aws-cdk-lib/aws-route53';
import { ApiGateway } from 'aws-cdk-lib/aws-route53-targets';
import { Construct } from 'constructs';

export interface ApiStackProps extends StackProps {
  hostedZoneId: string;
}

export class ApiStack extends Stack {
  api: RestApi;
  table: TableV2;
  private deployment!: Deployment;
  private apiDomainName: string;

  constructor(scope: Construct, id: string, private props: ApiStackProps) {
    super(scope, id, props);

    const apiDomainName = this.apiDomainName = 'api.example.matthewbonig.com';
    this.api = this.createApi();
    this.table = this.createTable();

    new CfnOutput(this, 'BaseApiGatewayURL', {
      value: `https://${this.api.restApiId}.execute-api.${this.region}.amazonaws.com/prod/`,
    });
    new CfnOutput(this, 'CustomURL', {
      value: `https://${apiDomainName}/v1/`,
    });

  }

  private createApi(): RestApi {
    const hostedZone = HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
      hostedZoneId: this.props.hostedZoneId,
      zoneName: 'example.matthewbonig.com',
    });
    const api = new RestApi(this, 'MyRestApi', {
      deploy: false,
      restApiName: 'CdkCourseApi',
      domainName: {
        domainName: this.apiDomainName,
        certificate: new Certificate(this, 'Certificate', {
          domainName: this.apiDomainName,
          validation: CertificateValidation.fromDns(hostedZone),
        }),
      },
    });

    new ARecord(this, 'ApiAliasRecord', {
      target: {
        aliasTarget: new ApiGateway(api),
      },
      recordName: 'api',
      zone: hostedZone,
    });

    api.root.addMethod('ANY');

    const deployment = this.deployment = new Deployment(this, 'Deployment', {
      api: api,
    });
    new Stage(this, 'Prod', {
      deployment,
      stageName: 'v1',
    });

    return api;
  }

  private createTable(): TableV2 {
    return new TableV2(this, 'MyTable', {
      partitionKey: {
        name: 'pk',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'sk',
        type: AttributeType.STRING,
      },
    });
  }

  public addToDeployment(resources: any[]) {
    for (const resource of resources) {
      this.deployment.addToLogicalId(resource);
    }
  }
}
