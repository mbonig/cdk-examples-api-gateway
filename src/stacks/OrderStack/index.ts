export * from './OrderStack';
export { OrderStackTests, OrderStackTestProps } from './OrderStackTests/OrderStackTests';

