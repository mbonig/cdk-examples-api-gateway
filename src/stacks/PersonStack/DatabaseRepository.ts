import { IdentifiableModel } from '../common/handlers/IdentifiableModel';

export interface IDatabaseRepository<ModelType extends IdentifiableModel> {
  get(id: string): Promise<ModelType>;

  getAll(): Promise<ModelType[]>;

  save(data: ModelType): Promise<ModelType>;
}
