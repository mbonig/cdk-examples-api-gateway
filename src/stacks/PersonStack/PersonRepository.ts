import { IDatabaseRepository } from './DatabaseRepository';
import { Person } from './Models';
import { IDatabaseAdapter, TableModel } from '../common/IDatabaseAdapter';


export interface IPersonRepository extends IDatabaseRepository<Person> {

}

export class PersonRepository implements IPersonRepository {
  constructor(private databaseAdapter: IDatabaseAdapter, private indexName: string) {

  }

  async getAll(): Promise<Person[]> {
    let people = await this.databaseAdapter.getAll(this.indexName);

    return people.map((person: TableModel) => ({
      id: person.sk,
      firstName: person.firstName,
      lastName: person.lastName,
    }));
  }

  async get(id: string): Promise<Person> {
    const person = await this.databaseAdapter.getItem(id, id);
    return { personId: person.sk, firstName: person.firstName, lastName: person.lastName };
  }

  async save(data: Person): Promise<Person> {
    const person: Person = {
      ...data,
    };
    await this.databaseAdapter.putItem({ pk: 'person', sk: person.personId, ...person });
    return person;
  }
}

