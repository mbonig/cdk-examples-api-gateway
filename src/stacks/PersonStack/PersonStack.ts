import { LambdaIntegration } from 'aws-cdk-lib/aws-apigateway';
import { AttributeType } from 'aws-cdk-lib/aws-dynamodb';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { INDEX_NAME, TABLE_NAME } from './PersonStack.GetPersonHandler';
import { RestApiResourceStack } from '../RestApiResourceStack';
import { RestResourceProps } from '../RestResourceStackProps';

export interface PersonStackProps extends RestResourceProps {
}

export class PersonStack extends RestApiResourceStack {

  public deploymentDependencies: any[] = [];

  constructor(scope: Construct, id: string, props: PersonStackProps) {
    super(scope, id, props);

    const indexName = 'people';
    const getHandler = new NodejsFunction(this, 'GetPersonHandler', {
      environment: {
        [TABLE_NAME]: props.table.tableName,
        [INDEX_NAME]: indexName,
      },
    });
    props.table.grantReadData(getHandler);

    const personResource = this.api.root.addResource('person', {});

    const getMethod = personResource.addMethod('GET', new LambdaIntegration(getHandler));

    const getByIdMethod = personResource
      .addResource('{id}', {})
      .addMethod('GET', new LambdaIntegration(getHandler));

    props.table.addGlobalSecondaryIndex({
      indexName: indexName,
      partitionKey: {
        name: 'personId',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'sk',
        type: AttributeType.STRING,
      },
    });

    this.deploymentDependencies.push({ method: getMethod.methodId }, { method: getByIdMethod.methodId });
  }

}
