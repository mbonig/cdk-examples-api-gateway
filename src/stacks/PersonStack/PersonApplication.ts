import { Person } from './Models';
import { IPersonRepository } from './PersonRepository';

export interface IPersonApplication {
  getById(id: string): Promise<Person>;

  getAll(): Promise<Person[]>;

}

export class PersonApplication implements IPersonApplication {
  constructor(private personRepository: IPersonRepository) {
  }


  async getById(id: string): Promise<Person> {
    if (!id) {
      throw new Error('Please provide an id');
    }
    return this.personRepository.get(id);
  }

  async getAll(): Promise<Person[]> {
    return this.personRepository.getAll();
  }

}
