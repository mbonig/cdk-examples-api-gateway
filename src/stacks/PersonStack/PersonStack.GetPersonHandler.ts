import { APIGatewayEvent } from 'aws-lambda';
import { PersonApplication } from './PersonApplication';
import { PersonRepository } from './PersonRepository';
import { DynamoDBAdapter } from '../common/IDatabaseAdapter';

export const TABLE_NAME = 'TABLE_NAME';
export const INDEX_NAME = 'INDEX_NAME';

export const handler = async (event: APIGatewayEvent) => {
  console.log(JSON.stringify(event, null, 2));
  const indexName = process.env[INDEX_NAME]!;
  const personApplication = new PersonApplication(
    new PersonRepository(
      new DynamoDBAdapter(process.env[TABLE_NAME]!),
      indexName,
    ),
  );

  if (event.pathParameters?.id) {
    const id = event.pathParameters.id;
    const person = await personApplication.getById(id);
    if (!person) {
      return {
        statusCode: 404,
        body: JSON.stringify({ message: 'Person not found' }),
      };
    }
    return {
      statusCode: 200,
      body: JSON.stringify(person),
    };
  }

  const persons = await personApplication.getAll();
  return {
    statusCode: 200,
    body: JSON.stringify(persons),
  };
};
