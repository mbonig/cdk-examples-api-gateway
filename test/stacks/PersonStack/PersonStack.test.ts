import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Code, CodeConfig } from 'aws-cdk-lib/aws-lambda';
import { ApiStack } from '../../../src/stacks/ApiStack';
import { PersonStack } from '../../../src/stacks/PersonStack';


let fromAssetMock: jest.SpyInstance;
beforeAll(() => {
  fromAssetMock = jest.spyOn(Code, 'fromAsset').mockReturnValue({
    isInline: false,
    bind: (): CodeConfig => {
      return {
        s3Location: {
          bucketName: 'my-bucket',
          objectKey: 'my-key',
        },
      };
    },
    bindToResource: () => {
      return;
    },
  } as any);
});

afterAll(() => {
  fromAssetMock?.mockRestore();
});

test('Snapshot', () => {
  const app = new App();
  const { api, table } = new ApiStack(app, 'Api', {});
  const stack = new PersonStack(app, 'test', {
    api: api,
    table: table,
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
