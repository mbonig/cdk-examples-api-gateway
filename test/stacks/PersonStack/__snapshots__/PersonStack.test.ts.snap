// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GetPersonHandler0EA48F33": {
      "DependsOn": [
        "GetPersonHandlerServiceRoleDefaultPolicy5615673A",
        "GetPersonHandlerServiceRole576B60AB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "my-bucket",
          "S3Key": "my-key",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "INDEX_NAME": "people",
            "TABLE_NAME": {
              "Fn::ImportValue": "Api:ExportsOutputRefMyTable794EDED1558A264E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "GetPersonHandlerServiceRole576B60AB",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetPersonHandlerServiceRole576B60AB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetPersonHandlerServiceRoleDefaultPolicy5615673A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::ImportValue": "Api:ExportsOutputFnGetAttMyTable794EDED1ArnDFAD2254",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetPersonHandlerServiceRoleDefaultPolicy5615673A",
        "Roles": [
          {
            "Ref": "GetPersonHandlerServiceRole576B60AB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RestApiperson1830F612": {
      "Properties": {
        "ParentId": {
          "Fn::ImportValue": "Api:ExportsOutputFnGetAttMyRestApi2D1F47A9RootResourceIdA8A99F67",
        },
        "PathPart": "person",
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApipersonGET904182CB": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "GetPersonHandler0EA48F33",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApiperson1830F612",
        },
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApipersonGETApiPermissionTesttestRestApi41EB688DGETperson6EB250F6": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetPersonHandler0EA48F33",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/test-invoke-stage/GET/person",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApipersonGETApiPermissiontestRestApi41EB688DGETperson18F55CAB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetPersonHandler0EA48F33",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/*/GET/person",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApipersonid7E2EA5B5": {
      "Properties": {
        "ParentId": {
          "Ref": "RestApiperson1830F612",
        },
        "PathPart": "{id}",
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApipersonidGET7C3772D5": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "GetPersonHandler0EA48F33",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApipersonid7E2EA5B5",
        },
        "RestApiId": {
          "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApipersonidGETApiPermissionTesttestRestApi41EB688DGETpersonid3E51EF8C": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetPersonHandler0EA48F33",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/test-invoke-stage/GET/person/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApipersonidGETApiPermissiontestRestApi41EB688DGETpersonid04966456": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "GetPersonHandler0EA48F33",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::ImportValue": "Api:ExportsOutputRefMyRestApi2D1F47A95DC88B97",
              },
              "/*/GET/person/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
